uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/WorkspaceProjectMigrationServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+import org.guvnor.structure.repositories.impl.git.GitRepository;
+import org.uberfire.backend.server.util.Paths;
+import org.uberfire.io.IOService;
+import org.uberfire.java.nio.file.Path;
+
+public class WorkspaceProjectMigrationServiceImpl
+        implements WorkspaceProjectMigrationService {
+
+    private WorkspaceProjectService workspaceProjectService;
+    private RepositoryService repositoryService;
+    private Event<NewProjectEvent> newProjectEvent;
+    private RepositoryCopier repositoryCopier;
+    private ModuleService<? extends Module> moduleService;
+    private IOService ioService;
+
+    public WorkspaceProjectMigrationServiceImpl() {
+    }
+
+    @Inject
+    public WorkspaceProjectMigrationServiceImpl(final WorkspaceProjectService workspaceProjectService,
+                                                final RepositoryService repositoryService,
+                                                final OrganizationalUnitService organizationalUnitService, // TODO remove unused
+                                                final Event<NewProjectEvent> newProjectEvent,
+                                                final RepositoryCopier repositoryCopier,
+                                                final ModuleService<? extends Module> moduleService,
+                                                final @Named("ioStrategy") IOService ioService) {
+        this.workspaceProjectService = workspaceProjectService;
+        this.repositoryService = repositoryService;
+        this.newProjectEvent = newProjectEvent;
+        this.repositoryCopier = repositoryCopier;
+        this.moduleService = moduleService;
+        this.ioService = ioService;
+    }
+
+    @Override
+    public void migrate(final WorkspaceProject legacyWorkspaceProject) {
+        new Migrator(legacyWorkspaceProject).migrate();
+    }
+
+    private class Migrator {
+
+        private final WorkspaceProject legacyWorkspaceProject;
+        private final Map<String, Repository> newRepositories = new HashMap<>();
+
+        public Migrator(final WorkspaceProject legacyWorkspaceProject) {
+            this.legacyWorkspaceProject = legacyWorkspaceProject;
+        }
+
+        public void migrate() {
+
+            copyModulesToRepositories();
+            fireNewProjectEvents();
+        }
+
+        private void fireNewProjectEvents() {
+            for (final Repository repository : newRepositories.values()) {
+                final WorkspaceProject newWorkspaceProject = workspaceProjectService.resolveProject(repository);
+                newProjectEvent.fire(new NewProjectEvent(newWorkspaceProject));
+            }
+        }
+
+        private void copyModulesToRepositories() {
+
+            for (final Branch branch : legacyWorkspaceProject.getRepository().getBranches()) {
+
+                for (final Module module : moduleService.getAllModules(branch)) {
+
+                    if (!newRepositories.containsKey(module.getModuleName())) {
```

Comment:
Here we're silently skipping creation of a repo. Shouldn't we at least log a warning ("Not creating BLA because BLA already contains BLA" (fill in the BLAs :)?

Reply:
So this is a tricky issue. We could add logging here, but it wouldn't do anything because this code is currently only called from the CLI migration tool, which deactivates all logging in order to hide console spam.

Ideally, we want the tool to help the user rename the module in conflict. That being said, the goal of the migration tool in this PR is to support the happy path, so I think we should open a JIRA for this enhancement (supporting module-name conflict resolution in the migration tool) and leave this for a subsequent PR.

Reply:
Ok, will file new jira for that.

Reply:
FYI: Jira filed https://issues.jboss.org/browse/AF-970